[develop] Add support for Capacity Blocks for ML

enrico-usai opened the pull request:
### Capacity Block for ML

Capacity Block reservation for Machine Learning (CB) allow AWS users to reserve blocks of GPU capacity for a future start date and for a time duration of their choice. They are capacity reservations of type `capacity-block`, with additional attributes like: `StartDate`, `EndDate` and additional states.

Capacity Block instances are managed by ParallelCluster as a static nodes but in a peculiar way. ParallelCluster permits to create a cluster even if the CB reservation is not yet active and automatically launch the instances when it will become active.

The Slurm nodes corresponding to compute resources associated to CB reservations are kept in maintenance until the CB start time is reached. When CB is not active, the slurm nodes will stay in a Slurm reservation/maintenance state, associated to the slurm admin user, it means that the nodes can accept jobs but they will stay in pending, until the reservation will be removed.

Clustermgtd will automatically create/delete slurm reservations, putting the related CB nodes in maintenance or not, accordingly to the CB state. The Slurm reservation will be removed when CB is active, nodes will start and become available for the pending jobs or for new jobs submissions.

When the CB end time is reached nodes will be re-inserted in a reservation/maintenance state. It’s up to the user to resubmit/requeue the jobs to a new queue/compute-resource when CB time is ended.

### CapacityBlockManager

`CapacityBlockManager` is a new class that will perform the following actions:
* when CB is not active or expired, a new Slurm reservation will be created/updated
* when CB is active, Slurm reservation will be removed and nodes will become standard static instances

Slurm reservation name will be `pcluster-{capacity_block_reservation_id}`

CapacityBlockManager will be initialized by clustermgtd with region, boto3_config and fleet_config info.
Fleet config is the existing `fleet-config.json` file, that has been extended to include capacity reservation ids and capacity type (ondemand vs spot vs capacity-block).
The manager will reload fleet-config info and capacity reservation info every time the daemon is restarted
or when the config is modified.
Current logic updates CB reservation info every 10 minutes.

The manager will remove the nodes associated with CB reservations that are not active
from the list of unhealthy static nodenames.

### CapacityBlock

`CapacityBlock` is a new class to store internal info about the capacity block,
merging them from ec2 (e.g. capacity block state) and from the config (i.e. list of nodes associated to it).

### Managing slurm reservations
Created a new set of commands (using scontrol) to manage the Slurm reservations:
* create a reservation
  * Use `slurm` as default user rather than root, given that "slurm" is the admin in ParallelCluster setup.
* update an existing reservation
  * I put in common steps to populate create and update commands since they are very similar.
* check if reservation exists
* delete_reservation
* added the reservation_name attribute to the `SlurmNode` class.

#### Leftover slurm reservations

When a CB is removed from a cluster config during a cluster update, we need to remove related slurm reservations.
We're retrieving all the slurm reservations from slurm and deleting them if they are no longer associated with existing CBs in the config.

### Nodes in maintenance state

After this patch the nodes in `MAINTENANCE+RESERVED` state will be excluded by the static nodes in replacement list.
Will see in the "replacement list" only nodes that are not yet up (as before) and nodes that are not in maintenance.

This mechanism permits the daemons (or the user) to avoid replacement of turned down static nodes (`DOWN`) by putting them in maintenance as preliminary step.
This works only for nodes in both `MAINTENANCE` and `RESERVED` state, nodes only in `RESERVED` state will be replaced as before.

In the future, we can extend this mechanism to have an additional field to check (e.g. skip them only if the maintenance is set by the `root` user).

### boto3 layer

Added boto3 layer, this code is taken from the CLI, removing the caching mechanism.
This permits to decouple node daemons code from boto3 calls, adding exceptions catching mechanism.

### Managing exceptions

Defined a new `SlurmCommandError` to identify errors coming from scontrol commands.
Defined a new `SlurmCommandErrorHandler` to permit to handle `SlurmCommandError` and log messages.

Added `retry (max-attempt:2, wait=1s)` and `SlurmCommandErrorHandler` decorator to all the reservations commands.

The main method of the `CapacityBlockManager` (`get_reserved_nodenames`), called by clustermgtd cannot raise any exception.
If the error in the `update_slurm_reservation` happens when updating a single capacity block/slurm reservation,
this is catched and the loop will continue, updating the others.

If there is a generic error like `AWSClientError` (wrapped to `CapacityBlockManagerError`) the entire list of `capacity_blocks` and `reserved_nodes` won't be changed, logging an error.

If `cleanup_leftover_slurm_reservation` fails this will be logged, but the process will continue.

### FleetManager changes

The main difference between on-demand, spot and capacity-block instances, is that capacity-block requires `Market-Type=capacity-block` but this is added to the Launch Template at generation time from the CLI/CDK.
Capacity reservation id and capacity type info for the compute resources are saved in the `config-fleet.json` file, that is generated by the cookbook according to Cluster configuration.

According to this file the node is able to understand if the compute-resource is associated to a CB reservation and add the following additional info:
```
        "OnDemandOptions": {
            ...
            "CapacityReservationOptions": {"UsageStrategy": "use-capacity-reservations-first"},
        },
        "TargetCapacitySpecification": {...."DefaultTargetCapacityType": "capacity-block"},
    }
```

The FleetManager has been modified to support empty `AllocationStrategy` because CB does not support any of the existing options.

### Tests
* The new and modified code is verified by unit tests.

### References
* Slurm reservations: https://slurm.schedmd.com/reservations.html
* Capacity Blocks: https://docs.aws.amazon.com/en_us/AWSEC2/latest/UserGuide/ec2-capacity-blocks.html
* CLI patch: https://github.com/aws/aws-parallelcluster/pull/5817
* Cookbook patch: https://github.com/aws/aws-parallelcluster-cookbook/pull/2526

### Logging

Slurm reservation creation
```
2023-11-07 12:43:55,361 - [slurm_plugin.capacity_block_manager:_retrieve_capacity_blocks_from_fleet_config] - INFO - Retrieving Capacity Blocks from fleet configuration.
2023-11-07 12:43:55,361 - [slurm_plugin.capacity_block_manager:_update_capacity_blocks_info_from_ec2] - INFO - Retrieving Capacity Blocks information from EC2 for cr-0296d8df657e57a7b
2023-11-07 12:43:55,384 - [aws.common:_log_boto3_calls] - INFO - Executing boto3 call: region=us-east-2, service=ec2, operation=DescribeCapacityReservations, params={'CapacityReservationIds': ['cr-0296d8df657e57a7b']}
2023-11-07 12:43:55,512 - [common.schedulers.slurm_reservation_commands:is_slurm_reservation] - INFO - Slurm reservation pcluster-cr-0296d8df657e57a7b not found.
2023-11-07 12:43:55,513 - [slurm_plugin.capacity_block_manager:_log_cb_info] - INFO - Capacity Block reservation cr-0296d8df657e57a7b is in state scheduled. Creating Slurm reservation pcluster-cr-0296d8df657e57a7b for nodes queue1-st-p5-1.
2023-11-07 12:43:55,513 - [common.schedulers.slurm_reservation_commands:create_slurm_reservation] - INFO - Creating Slurm reservation with command: sudo /opt/slurm/bin/scontrol create reservation
Reservation created: pcluster-cr-0296d8df657e57a7b
2023-11-07 12:43:55,573 - [slurm_plugin.clustermgtd:_find_unhealthy_slurm_nodes] - INFO - The nodes queue1-st-p5-1 are associated with unactive Capacity Blocks, they will not be considered as unhealthy nodes.
2023-11-07 12:43:55,574 - [slurm_plugin.slurm_resources:_is_static_node_ip_configuration_valid] - WARNING - Node state check: static node without nodeaddr set, node queue1-st-p5-1(queue1-st-p5-1), node state DOWN+CLOUD+NOT_RESPONDING+POWERING_UP:
2023-11-07 12:43:55,574 - [slurm_plugin.clustermgtd:_terminate_orphaned_instances] - INFO - Checking for orphaned instance
```

No time to update, `capacity_block_manager` does anything.
```
2023-11-07 17:30:51,134 - [slurm_plugin.clustermgtd:_maintain_nodes] - INFO - Following nodes are currently in replacement: (x0) []
2023-11-07 17:30:51,135 - [slurm_plugin.clustermgtd:_find_unhealthy_slurm_nodes] - INFO - The nodes associated with inactive Capacity Blocks and not considered as unhealthy nodes are: queue1-st-p5-1
2023-11-07 17:30:51,135 - [slurm_plugin.slurm_resources:_is_static_node_ip_configuration_valid] - WARNING - Node state check: static node without nodeaddr set, node queue1-st-p5-1(queue1-st-p5-1), node state IDLE+CLOUD+MAINTENANCE+POWERED_DOWN+RESERVED:
2023-11-07 17:30:51,135 - [slurm_plugin.clustermgtd:_terminate_orphaned_instances] - INFO - Checking for orphaned instance
```

Time to update, because update period passed (I set it to 2 minutes for testing the behaviour)
```
2023-11-07 17:31:51,163 - [slurm_plugin.clustermgtd:_maintain_nodes] - INFO - Following nodes are currently in replacement: (x0) []
2023-11-07 17:31:51,163 - [slurm_plugin.capacity_block_manager:_retrieve_capacity_blocks_from_fleet_config] - INFO - Retrieving Capacity Blocks from fleet configuration.
2023-11-07 17:31:51,163 - [slurm_plugin.capacity_block_manager:_update_capacity_blocks_info_from_ec2] - INFO - Retrieving Capacity Blocks information from EC2 for cr-0296d8df657e57a7b
2023-11-07 17:31:51,177 - [aws.common:_log_boto3_calls] - INFO - Executing boto3 call: region=us-east-2, service=ec2, operation=DescribeCapacityReservations, params={'CapacityReservationIds': ['cr-0296d8df657e57a7b']}
2023-11-07 17:31:51,310 - [slurm_plugin.capacity_block_manager:_log_cb_info] - INFO - Capacity Block reservation cr-0296d8df657e57a7b is in state scheduled. Nothing to do. Already existing Slurm reservation pcluster-cr-0296d8df657e57a7b for nodes queue1-st-p5-1.
2023-11-07 17:31:51,337 - [slurm_plugin.clustermgtd:_find_unhealthy_slurm_nodes] - INFO - The nodes associated with inactive Capacity Blocks and not considered as unhealthy nodes are: queue1-st-p5-1
2023-11-07 17:31:51,338 - [slurm_plugin.slurm_resources:_is_static_node_ip_configuration_valid] - WARNING - Node state check: static node without nodeaddr set, node queue1-st-p5-1(queue1-st-p5-1), node state IDLE+CLOUD+MAINTENANCE+POWERED_DOWN+RESERVED:
2023-11-07 17:31:51,338 - [slurm_plugin.clustermgtd:_terminate_orphaned_instances] - INFO - Checking for orphaned instance
```

Commit:
Extend scontrol output parser to be able to retrieve reservation info

After this patch the `_parse_nodes_info` method is able to retrieve `ReservationName` field.
This info will be saved in the `reservation_name` attribute of the SlurmNode.

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Add commands to manage slurm reservation

Add commands to
* create a reservation
  * Use slurm as default user rather than root,
    given that "slurm" is the admin in ParallelCluster setup.
* update an existing reservation
  * I put in common steps to populate create and update commands
    since they are very similar.
* check if reservation exists
* delete_reservation

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Add SlurmNode methods to check reservation state

Added:
* _is_reserved -> when there is RESERVED state
* is_in_maintenance -> when there are both RESERVED and MAINTENANCE states.

### Tests
* Added unit tests for the new methods.

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Fix Node class to use in mocks

The test is for static nodes, not dynamic.

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Exclude nodes in maintenance from static nodes to be replaced

After this patch the nodes in `MAINTENANCE+RESERVED` state
will be excluded by the static nodes in replacement list.

Will see in the "replacement list" only nodes that are not yet up (as before)
and nodes that are not in maintenance.

This mechanism permits the daemons (or the user) to avoid replacement
of turned down static nodes (`DOWN`) by putting them in maintenance as preliminary step.

This works only for nodes in both `MAINTENANCE` and `RESERVED` state,
nodes only in `RESERVED` state will be replaced as before.

In the future, we can extend this mechanism to have an additional field to check
(e.g. skip them only if the maintenance is set by the `root` user).

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Add boto3 layer

This code is taken from the CLI, removing the caching mechanism.

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Introduce CapacityBlockManager to handle instances part of a Capacity Block reservation

### Capacity Block reservation

Capacity Block reservation for ML (CB) allow AWS users to reserve blocks of GPU capacity for a future start date
and for a time duration of their choice.
They are capacity reservations of type `capacity-block`, with additional attributes like:
`StartDate`, `EndDate` and additional states.

Nodes part of the CB cannot be used until the CB time started.
Jobs assigned to these nodes will be kept in pending until the CB is active,
to achieve this we're going to use Slurm reservation, in particular we'll set the nodes
in maintenance state when CB is not yet active or when CB time is expired.

### CapacityBlockManager

CapacityBlockManager is a new class that will perform the following actions:
* when CB is not active or expired, a new Slurm reservation will be created/updated
* when CB is active, Slurm reservation will be removed and nodes will become standard static instances

Slurm reservation name will be "pcluster-{capacity_block_reservation_id}"

CapacityBlockManager will be initialized by clustermgtd with region, boto3_config and fleet_config info.
Fleet config is used to retrieve capacity reservation ids and type (ondemand vs capacity-block).
The manager will reload fleet-config info and capacity reservation info every time the daemon is restarted
or when the config is modified.
The logic updates capacity block reservation info every 10 minutes.

The manager will remove the nodes associated with capacity-block reservations that are not active
from the list of unhealty static nodenames.

### CapacityBlock

CapacityBlock is a new class to store internal info about the capacity block,
merging them from ec2 (e.g. capacity block state) and from the config (i.e. list of nodes associated to it)

### Leftover slurm reservations

When a CB is removed from a cluster config during a cluster update
we need to remove related slurm reservations.
We're checking slurm reservations for the given nodes and deleting them
when are managed by CapacityBlockManager but no longer in the config.

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Add command to get Slurm reservations info using scontrol

Defined a new dataclass to store Slurm reservation info.

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Cleanup all the leftover slurm reservations even if not related to given nodes

Previously the code was only deleting leftover slurm reservations
from the given nodes, but if the CB is moved from a queue to another
the list of nodes might not contain all the reservation.

With this new logic we're retrieving all the slurm reservations from slurm
and deleting them if they are no longer associated with existing CBs in the config.

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Rename does_slurm_reservation_exist to is_slurm_reservation

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Define _is_time_to_update_capacity_blocks_info method and fix internal logic

Return true if it's the moment to update capacity blocks info, from ec2 and from config.
This is true when the CapacityBlockManager is not yet initialized
`self._capacity_blocks_update_time == None` (this was missing before this patch) and every 10 minutes.

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Add robustness when updating info from ec2

Even if this should never happen, the code is now able to detect if we're trying
to update info from capacity blocks not in the internal map.

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Improve main method of CapacityBlockManager

Now the update of capacity_blocks info from config, from ec2 and from nodes
is done in the same method.

_update_slurm_reservation is now just updating reservations,
the collection of reserved nodes is done upstream.

Add unit tests to ensure the following items are correct:
- update time attribute
- internal capacity blocks list attribute
- reserved nodes are only updated if there are no internal errors and only include nodes from not active CBs.

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Minor improvements to debugging messages

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Move management of capacity blocks in a common place for static and dynamic nodes

Previously the logic was only applied to static nodes,
now the list of reserved nodes is evaluated for all the nodes.
Do not add reserved nodes to the list of all unhealthy nodes.

Added a new configuration parameter disable_capacity_blocks_management.

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Rename _capacity_blocks_from_config to _retrieve_capacity_blocks_from_fleet_config

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Avoid to raise an exception if there is an error updating a single capacity block

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Add essential logic to manage internal CapacityBlockManager exceptions

Now all the main logic is in a try/except block.
The manager is not raising any exception, but instead keeping the
previous value for the list of reserved nodenames.

Added logic to manage AWSClientError when contacting boto3 and
converting it to CapacityBlockManagerError.

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Manage internal errors when managing Capacity Blocks and related slurm reservations

Defined a new SlurmCommandError to identify errors coming from scontrol commands.
Defined a new SlurmCommandErrorHandler to permit to handle SlurmCommandError and log messages.

Added retry (max-attempt:2, wait=1s) and SlurmCommandErrorHandler decorator
to all the reservations commands.

Improved is_slurm_reservation command to be able to parse stderr and stdout
to retrieve reservation information.

Now the main method `get_reserved_nodenames`, called by clustermgtd cannot raise any exception.

If the error in the `update_slurm_reservation` happens when updating a single capacity block/slurm reservation,
this is catched and the loop will continue, updating the others.

If there is a generic error like AWSClientError (wrapped to CapacityBlockManagerError)
the entire list of capacity_blocks and reserved_nodes won't be changed, logging an error.

If cleanup_leftover_slurm_reservation fails this will be logged, but the process will continue.

Extended unit tests to cover command retries and error catching.

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Rename slurm_reservation_name_to_id to capacity_block_id_from_slurm_reservation_name

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Improve indentation

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Manage empty allocation strategy when using capacity block

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Add unit tests for fleet manager when using capacity block

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Rename parameters of is_healthy and is_state_healthy methods of SlurmNodes

From a SlurmNode perspective "terminate_down/drain_nodes" does not have any sense.
We can instead pass a flag to the is_healthy function to say if we want to consider
down/drain nodes as unhealthy.

Signed-off-by: Enrico Usai <usai@amazon.com>

enrico-usai commented on tests/common/schedulers/test_slurm_reservation_commands.py:
```diff
@@ -0,0 +1,514 @@
+# OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions and
+# limitations under the License.
+import logging
+import subprocess
```
Fixed.

Commit:
Improve unit tests, covering missing error cases

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Unify CapacityBlockReservationInfo with CapacityReservationInfo and add unit tests

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Add region to Boto3 client initialization

Region is required by describe-capacity-reservations call.

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Minor logging improvements

Do not log errors when checking for Slurm reservation existence
Do not log slurm commands

Signed-off-by: Enrico Usai <usai@amazon.com>

lukeseawalker commented on CHANGELOG.md:
```diff
@@ -7,6 +7,7 @@ This file is used to list changes made in each version of the aws-parallelcluste
 ------
 
 **ENHANCEMENTS**
+- Add support for Capacity Blocks for ML.
```
what about "EC2 Capacity Blocks"?

enrico-usai replied:
you're right, I'll change it.

enrico-usai replied:
modified.

Commit:
Fix ec2_client usage

It is an object, so it is not callable.
Fixing this I found that we were wrongly calling describe_capacity_reservations
by passing a dict rather than a list of ids.

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Fix slurm reservation creation

Start time is required and when adding a start time
the duration is required as well.

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Mark nodes as reserved only if Slurm reservation is correctly created

Previously the node was considered as reserved and removed from unhealthy nodes,
even if the slurm reservation process was failing.
Now we're checking if the slurm reservation has been correctly created/updated.

If the slurm reservation actions for ALL the CBs are failing,
we're not marking the nodes as reserved and not updating
the internal list of capacity blocks and update time.

If one of the reservation is updated correctly
the list of reserved nodes and update time attributes are updated.

Extended unit tests accordingly.

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Avoid to call update_slurm_reservation when it is not required

Previously we were calling scontrol command to update the reservation
every 10 minutes, but this is useless.

The only case where we need to update the reservation is to update the list
of nodes, this can happen only at cluster update time, so when Clustermgtd is restarted.
To identify this moment we are using the `_capacity_blocks_update_time` attribute.

Defined a new `_is_initialized` method and using it for `_update_slurm_reservation`

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Fix is_time_to_update_capacity_blocks_info method

Previously we were checking if `seconds_to_minutes()` was `True`.
Now we're correctly checking if `seconds_to_minutes` is `> CAPACITY_BLOCK_RESERVATION_UPDATE_PERIOD`

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Do not print warning messages for reserved nodes

In the log there was an entry for reserved instances saying:
```
WARNING - Node state check: static node without nodeaddr set, node queue1-st-p5-1(queue1-st-p5-1),
node state IDLE+CLOUD+MAINTENANCE+POWERED_DOWN+RESERVED
```

I'm removing the print for the reserved nodes.

Signed-off-by: Enrico Usai <usai@amazon.com>

lukeseawalker commented on tests/slurm_plugin/slurm_resources/test_slurm_resources.py:
```diff
@@ -287,6 +287,65 @@ def test_slurm_node_is_up(node, expected_output):
+            DynamicNode("queue1-dy-c5xlarge-1", "nodeip", "nodehostname", "IDLE+CLOUD+RESERVED", "queue1"),
+            True,
+        ),
+        (DynamicNode("queue1-dy-c5xlarge-1", "nodeip", "nodehostname", "RESERVED", "queue1"), True),
```
is it possible to have a state with MAINTENANCE but no RESERVED?

enrico-usai replied:
Nope, this is not possible, "maint" is a flag passed at reservation creation. It's not possible to set a node in maintenance without being in reserved state.

lukeseawalker commented on tests/slurm_plugin/test_capacity_block_manager.py:
```diff
@@ -0,0 +1,745 @@
+            (StaticNode("queue2-st-compute-resource1-4", "ip-1", "hostname-1", "some_state", "queue2"), False),
+        ],
+    )
+    def test_does_node_belong_to(self, capacity_block, node, expected_output):
```
Please add test parameter cases where 
* node is in CB queue but not in CB compute resource
* node is not ni CB queue and in a compute resource with same name of CB

enrico-usai replied:
Added.

lukeseawalker commented on tests/slurm_plugin/test_capacity_block_manager.py:
```diff
@@ -0,0 +1,745 @@
+                    ),
+                ],
+                [StaticNode("queue-cb-st-compute-resource-cb-1", "ip-1", "hostname-1", "some_state", "queue-cb")],
+                [True, True],
```
not clear to me why if there is an internal error, the slurm reservation create should succeed

enrico-usai replied:
Because it's an internal error  with the second capacity block cr-234567, that won't stop the loop. 
So the slurm reservation for the first CB will succeed.

lukeseawalker commented on tests/slurm_plugin/test_capacity_block_manager.py:
```diff
@@ -0,0 +1,745 @@
+                    )
+                ],
+                [StaticNode("queue-cb-st-compute-resource-cb-1", "ip-1", "hostname-1", "some_state", "queue-cb")],
+                [True, False],  # reservation creation failed for the first CB
```
do you mean for the second CB?

enrico-usai replied:
yes, sorry.

lukeseawalker commented on tests/slurm_plugin/test_capacity_block_manager.py:
```diff
@@ -0,0 +1,745 @@
+                ["node1"],
+            ),
+            (
+                # return nodenames accordingly to the state (only pending)
```
comment should be "only not active are returned", right?

enrico-usai replied:
yes, fixed.

lukeseawalker commented on tests/slurm_plugin/test_capacity_block_manager.py:
```diff
@@ -0,0 +1,745 @@
+        [
+            (
+                {
+                    "cr-123456": CapacityBlock("id", "queue-cb", "compute-resource-cb"),
```
* please add test case with no capacity blocks.
* can the same reservation ID be set on two (or more) different queues? How is this modeled? I suppose by two different CapacityBlock, please add the test case

enrico-usai replied:
This is a nice catch, thanks! 

Added a commit: https://github.com/aws/aws-parallelcluster-node/pull/591/commits/2a687fc9cad99b70d7e3932e8e978208fe31e9f6 to add support for multiple queues/compute-resources for the same CB.

Commit:
Support same CB in multiple queues and compute resources

Previously every CB was associated to a single queue and compute resource.
Now the queue/compute-resource association is stored in an internal map.

Signed-off-by: Enrico Usai <usai@amazon.com>

lukeseawalker commented on tests/slurm_plugin/test_capacity_block_manager.py:
```diff
@@ -0,0 +1,869 @@
+        [
+            ({}, "queue1", "cr1", {"queue1": {"cr1"}}),
+            ({"queue1": {"cr1"}}, "queue1", "cr2", {"queue1": {"cr1", "cr2"}}),
+            ({"queue1": {"cr1"}}, "queue2", "cr2", {"queue1": {"cr1"}, "queue2": {"cr2"}}),
```
please add following test case
`({"queue1": {"cr1"}}, "queue2", "cr1", {"queue1": {"cr1"}, "queue2": {"cr1"}}),`

enrico-usai replied:
added

enrico-usai committed:
Merge branch 'develop' into wip/cbr
